webofneeds/won-owner/src/main/java/won/owner/model/KeystorePasswordHolder.java

```diff
@@ -0,0 +1,46 @@
+package won.owner.model;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+import won.owner.service.impl.KeystorePasswordUtils;
+
+@Entity
+@Table(name = "keystore_password")
+public class KeystorePasswordHolder {
+	@Id
+	@GeneratedValue
+	@Column(name = "id")
+	private Long id;
+	
+	@Column(name= "encrypted_password")

```

Comment:
so we encrypt passwords? why are we not securely hashing them, or do we anyway do that?

Reply:
correct. It's easy to get confused though. Each user account has a keystore associated. We'll call the user's self-chosen password ´password` and the keystore password `keystorePassword`. The keystore password is a a secure random string, generated when the user account is created (or when it is first accessed and does not have one). The `keystorePassword` is saved in the database, (table `keystore_password`) but it is symmetrically encrypted with a key derived from the user's password in a reproducible manner. Thus, you can only unlock the keystore if you have the user's password in cleartext (which is only the case while we're processing the login request). The unlocked keystore and the keystore password are put into the user's autentication data, which is available within the session. So, user data is at risk during open sessions. (We can improve that later). If the user checks `remember Me`, the remember-me token is stored in the `persistent_logins` table (and the token value is set as the `remember-me` cookie). When that happens, another random string is generated with which the `keystorePassword` is symmetrically encrypted. The encrypted `keystorePassword` is stored in the `keystore_password` table. The persistent login entry is connected with that entry and a cookie named `won.unlock` is set with the key to unlock it. Thus, when the client authenticates using remember-me, it also has a cookie that is used to decrypt the `keystore_password`  and put it in the session.